### PR TITLE
NOX: catch std::runtime_error instead of char* for Trilinos 14.2.0 onwards

### PR DESCRIPTION
**Pull request by tamiko**
Fixes: https://cdash.dealii.org/test/254076

In reference to #15383

### CODE REVIEW

File: include/deal.II/trilinos/nox.templates.h [hunk 1]
```diff
@@ -1161,12 +1161,21 @@ namespace TrilinosWrappers
                         ExcNOXNoConvergence());
           }
       }
-    // See if NOX returned by triggering an exception. In a sign of generally
-    // poor software design, NOX throws an exception that is not of a class
-    // derived from std::exception, but just a char*. That's a nuisance -- you
-    // just have to know :-(
+    // See if NOX returned by triggering an exception.
+#if DEAL_II_TRILINOS_VERSION_GTE(4, 2, 0)
+    // Starting with Trilinos version 4.2.0 NOX started to throw a
```
Comment (kronbichler) [inline include/deal.II/trilinos/nox.templates.h:1166]: Do you mean 
```suggestion
#if DEAL_II_TRILINOS_VERSION_GTE(14, 2, 0)
    // Starting with Trilinos version 14.2.0 NOX started to throw a
```
Version 4.2 must be long before I started using Trilinos in 2007, I think the first version I used was 6 or 7.

Comment (tamiko) [reply include/deal.II/trilinos/nox.templates.h:1166]: :smile:

File: include/deal.II/trilinos/nox.templates.h [hunk 2]
```diff
@@ -1161,12 +1161,21 @@ namespace TrilinosWrappers
...
+      // See if NOX returned by triggering an exception.
+#    if DEAL_II_TRILINOS_VERSION_GTE(14, 2, 0)
+    // Starting with Trilinos version 14.2.0 NOX started to throw a
+    // std::runtime_error instead of a plain char*.
+    catch (const std::runtime_error &exc)
+      {
+        const char *s = exc.what();
+#    else
+    // In a sign of generally poor software design, NOX prior to Trilinos
+    // version 4.2.0 throws an exception that is not of a class derived
```
Comment (kronbichler) [inline include/deal.II/trilinos/nox.templates.h:1173]: One more:
```suggestion
    // version 14.2.0 throws an exception that is not of a class derived
```

**Review by kronbichler — APPROVED** [review]

**Review by bangerth — APPROVED** [review]
Ha, that's funny. I wrote the patch for Trilinos that changed the exception type :-)

**Pull request merged**